File: C++/Tree/BasicTree.cc
#include <iostream>
#include <string>
#include <vector>

struct Tree {
   std::vector<Tree> treeNodes;
   std::string data;

   Tree(std::string data);
   void addChild(Tree node);
   std::string printChild(int level);
};

Tree::Tree(std::string data) {
   this->data = data;
}

void Tree::addChild(Tree node) {
   treeNodes.push_back(node);
}

std::string Tree::printChild(int level) {
   std::string result;
   for (int i = 0; i < level; i++) {
      result += " ";
   }
   result += this->data + "\n";
   for (Tree node : treeNodes) {
      result += node.printChild(level + 1);
   }
   return result;
}

int main() {
   Tree drinks("Drinks");
   Tree hot("Hot");
   Tree cold("Cold");

   Tree tea("Tea");
   Tree coffee("Coffee");
   
   Tree cold_drink("Cold Drink");
   Tree fanta("Fanta");

   drinks.addChild(hot);
   drinks.addChild(cold);

   hot.addChild(tea);
   hot.addChild(coffee);
   
   cold.addChild(cold_drink);
   cold.addChild(fanta);
   std::cout << cold.printChild(2) << std::endl;
   return 0;
}

File: Python/Tree/BasicTree.py
class BasicTree:
    def __init__(self, data: str) -> None:
        self.__data = data
        self.__treeNodes = []

    def addChild(self, node) -> None:
        if node not in self.__treeNodes:
            self.__treeNodes.append(node)

    def printTree(self, level: int) -> str:
        space, result = "", ""

        for _ in range(level):
            space += " "

        result += (space + self.__data + "\n")

        for node in self.__treeNodes:
            result += node.printTree(level + 1)

        return result

drinks = BasicTree("Drinks")

hotDrinks = BasicTree("Hot Drinks")
coffee = BasicTree("Coffee")
tea = BasicTree("Tea")

hotDrinks.addChild(coffee)
hotDrinks.addChild(tea)

coldDrinks = BasicTree("Cold Drinks")
fanta = BasicTree("Fanta")
shake = BasicTree("Shake")

coldDrinks.addChild(fanta)
coldDrinks.addChild(shake)

drinks.addChild(hotDrinks)
drinks.addChild(coldDrinks)

print(drinks.printTree(0))
